src/video/rk.c

```diff
@@ -0,0 +1,519 @@
+/*
+ * This file is part of Moonlight Embedded.
+ *
+ * Copyright (C) 2015-2018 Iwan Timmer
+ * Copyright (C) 2018 Martin Cerveny, Daniel Mehrwald
+ *
+ * Moonlight is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Moonlight is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Moonlight; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <Limelight.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <assert.h>
+#include <time.h>
+#include <string.h>
+#include <errno.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/ioctl.h>
+#include <fcntl.h>
+#include <pthread.h>
+#include <unistd.h>
+
+#include <xf86drm.h>
+#include <xf86drmMode.h>
+#include <libdrm/drm_fourcc.h>
+
+#include <rockchip/rk_mpi.h>
+
+#define READ_BUF_SIZE (SZ_1M)
+#define MAX_FRAMES 16
+#define RK_H264 7
+#define RK_H265 16777220
+
+void *pkt_buf = NULL;
+pthread_t tid_frame, tid_display;
+drmModePlane *ovr = NULL;
+drmModeEncoder *encoder = NULL;
+drmModeConnector *connector = NULL;
+drmModeRes *resources = NULL;
+drmModePlaneRes *plane_resources = NULL;
+drmModeCrtcPtr crtc = {0};
+
+struct {
+    int fd;
+    uint32_t plane_id, crtc_id;
+    int frm_eos;
+    int crtc_width;
+    int crtc_height;
+    RK_U32 frm_width;
+    RK_U32 frm_height;
+    int fb_x, fb_y, fb_width, fb_height;
+    pthread_mutex_t mutex;
+    pthread_cond_t cond;
+    int fb_id;
+    int skipped_frames;
+} drm;
+
+struct {
+    MppCtx  ctx;
+    MppApi  *mpi;
+    MppPacket packet;
+    struct timespec first_frame_ts;
+    MppBufferGroup  frm_grp;
+    struct {
+        int prime_fd;
+        int fb_id;
+        uint32_t handle;
+    } frame_to_drm[MAX_FRAMES];
+} mpi;
+
+void *display_thread(void *param) {
+
+    int ret;
+    fprintf(stdout, "DISPLAY THREAD START\n");
+
+    while (!drm.frm_eos) {
+        int fb_id;
+
+        ret = pthread_mutex_lock(&drm.mutex);
+        assert(!ret);
+        while (drm.fb_id == 0) {
+            pthread_cond_wait(&drm.cond, &drm.mutex);
+            assert(!ret);
+            if (drm.fb_id == 0 && drm.frm_eos) {
+                ret = pthread_mutex_unlock(&drm.mutex);
+                assert(!ret);
+                fprintf(stdout, "DISPLAY THREAD END\n");
+                return NULL;
+            }
+        }
+        fb_id = drm.fb_id;
+        if (drm.skipped_frames) {
+            fprintf(stdout, "DISPLAY skipped %d\n", drm.skipped_frames);
```

Comment:
I don't think creating console output for every skipped frame is necessary, and the { and } can be removed